Add http.rb instrumentation to NewRelic

tjsousa opened the pull request:
Since Talkdesk/contacts-api-ruby uses `HalClient` that depends on http.rb (the gem!) and this HTTP client is not supported by the official NewRelic RPM, this optional gem can be included in projects using the client (or in the gem itself) to add external services monitoring in NewRelic for our Contacts API service.

Commit:
Initial version of http.rb instrumentation for NewRelic

Commit:
Allow gem to be included using require "newrelic/httprb"

Commit:
Add specs to http.rb instrumentation

axfcampos commented on lib/newrelic_httprb.rb:
```diff
@@ -0,0 +1,2 @@
+require "newrelic_httprb/instrumentation"
+require "newrelic_httprb/version"
```
💅  missing newline

jcmfernandes replied:
C'mon @ivoanjo, how come that you have let this lack of new line slip? ... 😆

jcmfernandes replied:
Seems like it got fixed in 60b61cbe9156646dfc0967662f03e3046867e441 😸 

Yes yes, I'm just here because I missed reviewing your code guys 😝 ❤️

axfcampos replied:
@jcmfernandes 🙌 😚

ivoanjo replied:
@jcmfernandes I was baiting you, and it worked ;D

ivoanjo commented:
Would be cool to add a license and publish this on rubygems (hint hint ;D).

ivoanjo commented on lib/newrelic_httprb/instrumentation.rb:
```diff
@@ -0,0 +1,33 @@
+
+        ::NewRelic::Agent::CrossAppTracing.tl_trace_http_request(wrapped_request) do
+          # RUBY-1244 Disable further tracing in request to avoid double
+          # counting if connection wasn't started (which calls request again).
```
This comment seems to be inspired by https://github.com/newrelic/rpm/blob/master/lib/new_relic/agent/instrumentation/net.rb#L24-L28 .

I'm not sure if this still applies to `http.rb`. The newrelic bug tracker is not public, but their specs that mention RUBY-1244 at https://github.com/newrelic/rpm/blob/master/test/new_relic/http_client_test_cases.rb seem to be specific to `NET::HTTP`.

Even if this applies, shouldn't we use the `NewRelic::Agent.disable_all_tracing` that they do in their code?

tjsousa replied:
Indeed, sorry for not signaling it but the branch was still a WIP. I copied this as a TODO to check if `http.rb` was subject to the same types of issues.

tjsousa replied:
I believe the `http_client_test_cases` is a generic suite of tests for the several client instrumentations, as hinted here:

https://github.com/newrelic/rpm/blob/master/test/multiverse/suites/typhoeus/typhoeus_test.rb#L7

However, I wasn't able to require the test cases from within my gem, so I am not really sure `http.rb` is subject to the same issue worked around in `RUBY-1244`. Any thoughts on how to proceed on this one?

ivoanjo replied:
Hmm, you're right. Maybe the solution here would be to fork newrelic, and add the testing code on the fork, and test it as part of their harness. It's not great, but we do get to take advantage of the many testcases they have already written for strange corner cases.

tjsousa replied:
I'm planning on doing so and contribute a PR to their gem as well. However, running on a forked version for our systems makes me nervous. Maybe we can use it to see if the implementation passes on their test harness and use this gem anyway? 

Most of the contrib gems for newrelic that we use don't have _any_ tests...

ivoanjo replied:
Yeah, that's what I was suggesting, to use the fork just as a workaround to allow us to test using their specs, nothing more :+1:

ivoanjo commented on lib/newrelic_httprb/instrumentation.rb:
```diff
@@ -0,0 +1,33 @@
+          ::NewRelicHTTP::HTTPResponse.new(response)
+        end
+
+        response
```
Not sure if this will work. Don't you need to initialize `response` before using it in the block? Otherwise the `response` in the block will be local to the block itself.

tjsousa replied:
![me and @axfcampos](https://encrypted-tbn3.gstatic.com/images?q=tbn:ANd9GcTD6U_b5Eh0qy7NyftAxzYMIFT41h9YY-c-6xJKpe-wgd_yKndu4lOsCW5P)

tjsousa replied:
We were testing the code and came across this same problem.

tjsousa replied:
Fixed in #2

ivoanjo commented on newrelic_httprb.gemspec:
```diff
@@ -0,0 +1,23 @@
+  gem.files         = `git ls-files`.split($/)
+  gem.test_files    = gem.files.grep(%r{^(test|spec|features)/})
+
+  gem.add_dependency 'newrelic_rpm', '~> 3.11'
```
Hmmm... curious. Any reason for 3.11 in particular?

tjsousa replied:
It's the minimum `newrelic_rpm` dependency that will work with latest versions of contrib gems we use:

[newrelic_moped](https://github.com/stevebartholomew/newrelic_moped/blob/master/newrelic_moped.gemspec#L16): `~> 3.11`
[padrino-rpm](https://github.com/Asquera/padrino-rpm/blob/master/padrino-rpm.gemspec): `~> 3.3.3`
[newrelic-redis](https://github.com/evanphx/newrelic-redis/blob/master/newrelic-redis.gemspec#L27): `~> 3.11`
[newrelic-zookeeper](https://github.com/jcmfernandes/newrelic-zookeeper/blob/master/newrelic-zookeeper.gemspec#L38): `~> 3.11`

This will also mean some systems (looking at you CTI) will have to update their NewRelic version.

Makes sense?

ivoanjo commented on lib/newrelic_httprb/wrappers.rb:
```diff
@@ -0,0 +1,51 @@
+    def to_hash
+      response.headers
+    end
+  end
```
Is this a copy of `NewRelic::Agent::HTTPClients::HTTPClientResponse`? I don't think their license allows this usage. Alternatively, we can just do

``` ruby
module NewRelicHTTP
  HTTPResponse = NewRelic::Agent::HTTPClients::HTTPClientResponse
```

...I think.

tjsousa replied:
Using the class equality was breaking the tests in the NewRelic RPM test suite for a reason I wasn't able to track. Due to a slightly different behaviour of `HTTP` response class being wrapped here, this is now slightly different from `HTTPClientResponse` so I think no copyright issues will arise.

Commit:
Fix incorrect require path

Commit:
Fix variable initialization

tjsousa commented on lib/newrelic_httprb/wrappers.rb:
```diff
@@ -0,0 +1,51 @@
+    def to_hash
+      response.headers
+    end
+  end
```
Good catch. It happens to be the same implementation as the wrapped class responds to the same `#headers` method. I will follow your suggestion to avoid licensing issues.

Commit:
Use an alias to define HTTPResponse to avoid copying the implementation

There may be licensing issues by using NewRelic's source code ipsis
verbis. This approach ties us to the existing implementation but as both
interfaces should be stable (implement #headers) we should be ok.

Commit:
Update README with license information

tjsousa committed:
Merge pull request #2 from axfcampos/fix-small-bugs

Fix small bugs

Commit:
Backport fixes of running instrumentation with the NewRelic test harness

After porting the instrumentation to be merged inside the NewRelic rpm
it was possible to run already existing functional testing for HTTP
client instrumentation which uncovered some failed expectations.

This commit backports those fixes.

See https://github.com/newrelic/rpm/pull/249

picaoao commented on lib/newrelic_httprb/wrappers.rb:
```diff
@@ -1,24 +1,5 @@
-      response.headers
-    end
-  end
+  HTTPResponse = NewRelic::Agent::HTTPClients::HTTPClientResponse
```
If we depend on that specific interface, despite it being assumed it is stable, I would say we can add some tests to make sure it is/will be as we are expecting. I'm not sure if this is already implicit in the existing tests. If it is, you can discard my comment 😄
